### ISSUE TITLE
Override if there is an existing jobId

### ISSUE THREAD

**Issue by JoHuang** [taskforcesh/bullmq#1192]
Is it possible to change the behavior of jobId?
In my cases, override instead of ignoring is required.
Thanks

**Comment by manast** [comment]
This is an interesting feature but needs to be thought about carefully, there may be some edge cases. For example, how to deal if the job has already started or has already been completed/failed? You probably do not want to override in these cases right?

**Comment by tolgap** [comment]
Is it at least possible to ignore jobs that were completed? It's handy for debugging and retrying. So when a job succeeds, but I want to retry it for whatever reason during debugging, I can't reuse the jobId for something that was already completed.

The solution would be to remove the job on completion, but I like to have a "papertrail" of 5000 jobs in a queue. So I usually do `removeOnComplete: 5000`.

**Comment by JoHuang** [comment]
Let me explain the use case.

It's a turn-based game, and every player has their action time (like 30s).
The game has a unique Id which I would use it as JobId.
If it's now Player A's turn, I need to set a delayed job at 30s from now to timed out the player if no action from him.
But mostly Player A takes an action so I have to set another delayed job at 30s from that time for the next player, and the original delayed job for Player A should not be fired. That's why I want to have "override" behavior  with the unique Id to keep there is always a delayed job for this game.

@manast  _how to deal if the job has already started or has already been completed/failed?_ 
In my case I just need to ignore the original job whatever the state it is.

**Comment by hyperair** [comment]
Hello, I'd like to chime in with another usecase -- we have something along the lines of jobs that rebuild a specific resource (specified by ID), and schedule jobs with `jobId` ``` `build:${resourceType}:${resourceId}` ``` in response to an event (like one of its dependencies being updated, e.g. like in a Makefile).

This works well and good for coalescing multiple requests so we don't duplicate calls, but it doesn't take into account cases where:

1. The rebuild job has started
2. While the rebuild job is running, one of the dependencies gets updated, and triggers a rebuild again -- BullMQ ignores this because the job already exists and is `active`
3. The rebuild job ends, but doesn't take into account the update that happened in step 2, so that update doesn't actually get synced for a long time (until something else triggers a rebuild of the resource again).

Ideally, in step 2, since the rebuild job is idempotent, we'd like to schedule the job to run again after step 3 to ensure that the resource is really up to date.

For this purpose, it would be great if BullMQ's `queue.add()` also supported something like the `job_key_mode=replace` behaviour of graphile-worker (described [here](https://github.com/graphile/worker#replacing-updating-and-removing-jobs)), as opposed to `job_key_mode=unsafe_dedupe` which is what BullMQ's current behaviour is, perhaps controlled using a flag in `JobsOptions`

**Comment by manast** [comment]
@hyperair wouldn't the "default" replace mode be the same as not specifying a custom Id?

**Comment by hyperair** [comment]
@manast By `default`, do you mean graphile-worker's `job_key_mode=replace`? In that case there are some differences from not specifying an ID:
- if the job is `waiting` or `delayed`, the second and third `queue.add` without specifying an ID would result in two or more identical duplicate jobs in the `waiting` or `delayed` states, but I want there to only be one, i.e. I still want some deduplication to happen.
- if the job is in `active`, then the job has already started, and I'd like there to be another job scheduled to ensure that the resource is up-to-date once everything is done.

In a nutshell, I just need at least one job execution to start *after* the final `queue.add()`

**Comment by manast** [comment]
The problem is that we cannot have more than 1 job with a given ID, so if the job is active we cannot add a new one. If the job is not active, then I think it would be the same as just updating the job's data: https://api.docs.bullmq.io/classes/Job.html#update

**Comment by manast** [comment]
Maybe you can detect if the job's data has been changed since the job started and if so process it again?

**Comment by hyperair** [comment]
> The problem is that we cannot have more than 1 job with a given ID, so if the job is active we cannot add a new one. If the job is not active, then I think it would be the same as just updating the job's data: https://api.docs.bullmq.io/classes/Job.html#update

There's a bit more than just updating the job's data that we need, and we ended up with a bunch of workarounds:
 - if the job is `delayed,` we want the latest `queue.add(...)`'s delay to take effect -- `job.changeDelay` takes care of this case. It's not important for us to kick a job further into the future, but we need to be able to reschedule it nearer to the present if it's already scheduled far into the future.
 - if the job is `completed`/`failed`, we want the job to get reprocessed, even if it failed before (maybe it'll work this time :sweat_smile:)
 - if the job is `waiting,` ideally we'd just like the latest `queue.add(...)`'s parameters to take effect, but it turns out it's really expensive on redis CPU to remove things from the `waiting` state if there's a large backlog, so we just update the data and leave it in `waiting`
 - if the job is `active`, that's where it gets really hacky, because we're just kicking it into the future using `job.moveToDelayed` without a token, which then causes the `active` job to fail with either `Missing lock` or `Job not in active state`

> Maybe you can detect if the job's data has been changed since the job started and if so process it again?

That's an interesting idea, I'll look into that, thanks.

**Comment by manast** [comment]
I think that most of the cases exposes in this thread can be resolved now either by using the standard behaviour of ignoring jobs with the same id or by using the deduplication options: https://docs.bullmq.io/guide/jobs/deduplication
